File: core/src/passes/global_cse.cpp
// core/src/passes/global_cse.cpp — F2-PS-004 Global Common Subexpression
// Elimination over an `ir::Function`.
//
// Forwards the intra-block CSE's available-expression hash table along
// dominator-tree edges where the dominated child has exactly one
// predecessor in the CFG, and that predecessor is its immediate
// dominator. Diamond / join blocks (multiple predecessors) start with
// an empty entry table — see `prisma/passes.hpp` for the soundness
// discussion and the deferred follow-up (classical available-expressions
// dataflow) that would tighten this.
//
// Within each block the algorithm is identical to
// `common_subexpression_eliminate` in `cse.cpp`. We re-implement
// rather than share a helper because the per-block state lifecycle
// here (entry table from idom, end table cached for descendants)
// makes the loop body different in subtle ways.

#include "prisma/passes.hpp"

#include <tuple>
#include <unordered_map>
#include <variant>

#include "prisma/dominators.hpp"

namespace prisma::passes {

namespace {

struct BinOpKey {
    ir::BinOpKind op;
    ir::Ref       lhs;
    ir::Ref       rhs;
    ir::OpSize    size;

    bool operator==(const BinOpKey& o) const noexcept = default;
};

struct BinOpKeyHash {
    std::size_t operator()(const BinOpKey& k) const noexcept {
        std::uint64_t h = 0xcbf29ce484222325ULL;
        const auto mix = [&](std::uint64_t v) {
            h ^= v;
            h *= 0x100000001b3ULL;
        };
        mix(static_cast<std::uint64_t>(k.op));
        mix(k.lhs);
        mix(k.rhs);
        mix(static_cast<std::uint64_t>(k.size));
        return static_cast<std::size_t>(h);
    }
};

using AvailTable = std::unordered_map<BinOpKey, ir::Ref, BinOpKeyHash>;

bool is_flushing_op(const ir::Op& op) noexcept {
    return std::visit([](auto const& x) -> bool {
        using T = std::decay_t<decltype(x)>;
        return std::is_same_v<T, ir::StoreReg>
            || std::is_same_v<T, ir::StoreMem>
            || std::is_same_v<T, ir::StoreMemTSO>
            || std::is_same_v<T, ir::CmpFlags>
            || std::is_same_v<T, ir::AluFlags>
            || std::is_same_v<T, ir::Syscall>;
    }, op);
}

}  // namespace

ir::Function global_cse(const ir::Function& fn) {
    if (fn.blocks.empty()) return fn;

    // id → position in `fn.blocks`. `idoms[idx[id]]` then yields the
    // immediate dominator's block ID.
    std::unordered_map<std::uint32_t, std::size_t> idx;
    idx.reserve(fn.blocks.size());
    for (std::size_t i = 0; i < fn.blocks.size(); ++i) {
        idx[fn.blocks[i].id] = i;
    }

    // Predecessor map derived from the public `successors()` API. Each
    // block id maps to the (possibly empty) list of predecessor ids.
    std::unordered_map<std::uint32_t, std::vector<std::uint32_t>> preds;
    for (const auto& b : fn.blocks) {
        for (auto s : ir::successors(fn, b.id)) {
            preds[s].push_back(b.id);
        }
    }

    const auto idoms = ir::dominators(fn);
    if (idx.find(fn.entry) == idx.end()) return fn;

    // Reverse postorder visitation: every block is visited after all of
    // its dominators (this is the standard CHK property). Unreachable
    // blocks don't appear and are left untouched in the output.
    const auto po = ir::postorder(fn);
    std::vector<std::uint32_t> rpo(po.rbegin(), po.rend());

    std::unordered_map<std::uint32_t, AvailTable> end_table;

    ir::Function out = fn;

    for (auto block_id : rpo) {
        AvailTable seen;
        if (block_id != fn.entry) {
            const auto pit = preds.find(block_id);
            if (pit != preds.end() && pit->second.size() == 1) {
                const std::uint32_t pred = pit->second.front();
                if (pred == idoms[idx[block_id]]) {
                    const auto eit = end_table.find(pred);
                    if (eit != end_table.end()) {
                        seen = eit->second;
                    }
                }
            }
        }

        auto& block = out.blocks[idx[block_id]];
        for (auto& stmt : block.stmts) {
            if (is_flushing_op(stmt.op)) {
                seen.clear();
                continue;
            }
            if (!std::holds_alternative<ir::BinOp>(stmt.op) || !stmt.result) {
                continue;
            }
            const auto& b = std::get<ir::BinOp>(stmt.op);
            const BinOpKey key{b.op, b.lhs, b.rhs, b.size};
            const auto it = seen.find(key);
            if (it != seen.end()) {
                stmt.op = ir::Op{ir::BinOp{
                    ir::BinOpKind::Or, it->second, it->second, b.size}};
            } else {
                seen[key] = *stmt.result;
            }
        }

        end_table[block_id] = std::move(seen);
    }

    return out;
}

}  // namespace prisma::passes

File: shell/prisma-passes/src/global_cse.rs
//! Global common subexpression elimination over a `Function`.
//!
//! Mirrors C++ `global_cse`. Forwards the intra-block available-expression
//! table along dominator-tree edges where the dominated child has exactly one
//! CFG predecessor and that predecessor is its immediate dominator. Join
//! blocks (multiple predecessors) start with an empty table — conservative but
//! sound. Within each block the logic matches `cse`.

use std::collections::HashMap;

use prisma_ir::{BinOpKind, Function, Op, OpSize, Ref};

use crate::cfg;
use crate::Pass;

type BinOpKey = (BinOpKind, Ref, Ref, OpSize);
type AvailTable = HashMap<BinOpKey, Ref>;

fn is_flushing_op(op: &Op) -> bool {
    matches!(
        op,
        Op::StoreReg(_)
            | Op::StoreMem(_)
            | Op::StoreMemTSO(_)
            | Op::CmpFlags(_)
            | Op::AluFlags(_)
            | Op::WriteFlagsPopcnt(_)
            | Op::WriteFlagsCountZero(_)
            | Op::Syscall(_)
    )
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub struct GlobalCse;

impl GlobalCse {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl Pass for GlobalCse {
    fn name(&self) -> &'static str {
        "global_cse"
    }

    fn run(&self, func: Function) -> Function {
        global_cse(func)
    }
}

/// Dominator-tree-forwarded CSE over the whole function.
pub fn global_cse(mut func: Function) -> Function {
    if func.blocks.is_empty() {
        return func;
    }

    // id -> position in func.blocks
    let idx: HashMap<u32, usize> = func
        .blocks
        .iter()
        .enumerate()
        .map(|(i, b)| (b.id, i))
        .collect();

    // Predecessor map from successors().
    let mut preds: HashMap<u32, Vec<u32>> = HashMap::new();
    for b in &func.blocks {
        for s in cfg::successors(&func, b.id) {
            preds.entry(s).or_default().push(b.id);
        }
    }

    let idoms = cfg::dominators(&func);
    if !idx.contains_key(&func.entry) {
        return func;
    }

    // Reverse postorder: each block visited after its dominators.
    let mut rpo = cfg::postorder(&func);
    rpo.reverse();

    let mut end_table: HashMap<u32, AvailTable> = HashMap::new();

    for block_id in rpo {
        let mut seen: AvailTable = AvailTable::new();
        if block_id != func.entry {
            if let Some(bpreds) = preds.get(&block_id) {
                if bpreds.len() == 1 {
                    let pred = bpreds[0];
                    if idoms.get(&block_id) == Some(&pred) {
                        if let Some(t) = end_table.get(&pred) {
                            seen.clone_from(t);
                        }
                    }
                }
            }
        }

        let Some(&block_pos) = idx.get(&block_id) else {
            continue; // defensive: postorder only yields real blocks, but guard anyway
        };
        let block = &mut func.blocks[block_pos];
        for stmt in &mut block.stmts {
            if is_flushing_op(&stmt.op) {
                seen.clear();
                continue;
            }
            if let (Some(result), Op::BinOp(b)) = (stmt.result, &stmt.op) {
                let key = (b.op, b.lhs, b.rhs, b.size);
                if let Some(&prev) = seen.get(&key) {
                    let size = b.size;
                    stmt.op = Op::BinOp(prisma_ir::BinOp {
                        op: BinOpKind::Or,
                        lhs: prev,
                        rhs: prev,
                        size,
                    });
                } else {
                    seen.insert(key, result);
                }
            }
        }

        end_table.insert(block_id, seen);
    }

    func
}

#[cfg(test)]
mod tests {
    use super::*;
    use prisma_ir::{BasicBlock, BinOp, Jump, Stmt};

    fn binop(result: u32, lhs: u32, rhs: u32) -> Stmt {
        Stmt::new(
            Some(result),
            Op::BinOp(BinOp {
                op: BinOpKind::Add,
                lhs,
                rhs,
                size: OpSize::I64,
            }),
        )
    }

    #[test]
    fn forwards_across_single_pred_idom_edge() {
        // block0: r2 = add r0,r1 ; jump 1
        // block1: r3 = add r0,r1  -> rewritten to Or r2,r2 (forwarded)
        let f = Function {
            entry: 0,
            blocks: vec![
                BasicBlock {
                    id: 0,
                    stmts: vec![
                        binop(2, 0, 1),
                        Stmt::new(None, Op::Jump(Jump { target_block: 1 })),
                    ],
                },
                BasicBlock {
                    id: 1,
                    stmts: vec![binop(3, 0, 1)],
                },
            ],
        };
        let out = global_cse(f);
        match &out.blocks[1].stmts[0].op {
            Op::BinOp(b) => {
                assert_eq!(b.op, BinOpKind::Or);
                assert_eq!((b.lhs, b.rhs), (2, 2));
            }
            other => panic!("expected forwarded copy, got {other:?}"),
        }
    }

    #[test]
    fn join_block_does_not_forward() {
        // 0 -> {1,2} -> 3. block3 has two preds, so no forwarding even if
        // both branches compute the same expr.
        let f = Function {
            entry: 0,
            blocks: vec![
                BasicBlock {
                    id: 0,
                    stmts: vec![
                        binop(2, 0, 1),
                        Stmt::new(
                            None,
                            Op::CondJump(prisma_ir::CondJump {
                                cond: 2,
                                if_true: 1,
                                if_false: 2,
                            }),
                        ),
                    ],
                },
                BasicBlock {
                    id: 1,
                    stmts: vec![Stmt::new(None, Op::Jump(Jump { target_block: 3 }))],
                },
                BasicBlock {
                    id: 2,
                    stmts: vec![Stmt::new(None, Op::Jump(Jump { target_block: 3 }))],
                },
                BasicBlock {
                    id: 3,
                    stmts: vec![binop(5, 0, 1)],
                },
            ],
        };
        let out = global_cse(f);
        // block3's add is NOT forwarded (join block starts empty).
        match &out.blocks[3].stmts[0].op {
            Op::BinOp(b) => assert_eq!(b.op, BinOpKind::Add),
            other => panic!("unexpected {other:?}"),
        }
    }

    #[test]
    fn intra_block_dedup_still_works() {
        let f = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![binop(2, 0, 1), binop(3, 0, 1)],
            }],
        };
        let out = global_cse(f);
        match &out.blocks[0].stmts[1].op {
            Op::BinOp(b) => assert_eq!((b.op, b.lhs, b.rhs), (BinOpKind::Or, 2, 2)),
            other => panic!("unexpected {other:?}"),
        }
    }

    #[test]
    fn empty_function_unchanged() {
        let f = Function {
            entry: 0,
            blocks: vec![],
        };
        assert_eq!(global_cse(f.clone()), f);
    }
}
